trieste/bayesian_optimizer.py

```diff
@@ -788,152 +822,152 @@ def _write_summary_init(
+            observation_plot_df["pareto"] = non_dominated(datasets[tag].observations)[1] == 0
+            observation_plot_df["observation type"] = observation_plot_df.apply(
+                lambda x: x["observations"] + x["pareto"] * " (non-dominated)",
+                axis=1,
+            )
+            pairplot = sns.pairplot(
+                observation_plot_df,
+                vars=columns,
+                hue="observation type",
+                hue_order=[
+                    "initial",
+                    "old",
+                    "new",
+                    "initial (non-dominated)",
+                    "old (non-dominated)",
+                    "new (non-dominated)",
+                ],
+                palette={
+                    "initial": "tab:green",
+                    "old": "tab:green",
+                    "new": "tab:orange",
+                    "initial (non-dominated)": "tab:purple",
+                    "old (non-dominated)": "tab:purple",
+                    "new (non-dominated)": "tab:red",
+                },
+                markers={
+                    "initial": "X",
+                    "old": "o",
+                    "new": "o",
+                    "initial (non-dominated)": "X",
+                    "old (non-dominated)": "o",
+                    "new (non-dominated)": "o",
+                },
+            )
+            logging.pyplot(f"{tag}.observations/_pairplot", pairplot.fig)
+            observation_plot_df.loc[
+                observation_plot_df["observations"] == "new", "observations"
+            ] = "old"
+            observation_plot_dfs[tag] = observation_plot_df
+
+    logging.scalar(
+        "wallclock/model_fitting",
+        model_fitting_timer.time,
+    )
+
+
+def write_summary_query_points(
+    datasets: Mapping[str, Dataset],
+    models: Mapping[str, TrainableProbabilisticModel],
+    search_space: SearchSpace,
+    query_points: TensorType,
+    query_point_generation_timer: Timer,
+    query_plot_dfs: MutableMapping[int, pd.DataFrame],
+) -> None:
+    """Write TensorBoard summary for the current step query points."""
+
+    if tf.rank(query_points) == 2:
+        for i in tf.range(tf.shape(query_points)[1]):
+            if len(query_points) == 1:
+                logging.scalar(f"query_points/[{i}]", float(query_points[0, i]))
+            else:
+                logging.histogram(f"query_points/[{i}]", query_points[:, i])
+        logging.histogram("query_points/euclidean_distances", lambda: pdist(query_points))
+
+    if pd and sns and logging.include_summary("query_points/_pairplot"):
+        columns = [f"x{i}" for i in range(tf.shape(query_points)[1])]
+        qp_preds = query_points
+        for tag in datasets:
+            pred = models[tag].predict(query_points)[0]
+            qp_preds = tf.concat(
+                [qp_preds, tf.cast(pred, query_points.dtype)], 1
+            )
+            if tf.shape(pred)[-1] == 1:
```

Comment:
here you have two cases: either "pred" is one-dimensional, then you use the tag directly, or it is multidimensional (typically, a ModelStack with tag OBJECTIVE), then you need more column names, e.g. OBJECTIVE1, OBJECTIVE2

Reply:
might be tidier using `output_dim = tf.shape(tagged_output[tag].observations)[-1]`